core: Change core errors to the correct formatting

majaziri opened the pull request:
closes #3192

flomonster reviewed (changes requested):
Can you rebase with only your commits?

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/OSRDError.java:
```diff
@@ -35,15 +36,20 @@ public abstract class OSRDError extends RuntimeException {
+     * Execution context of the error 
+     * for example Signal id, train Id, etc
+     */
+    public final Map<String, Object> context = new HashMap<>();
```
```suggestion
    public final Map<String, ErrorContext> context = new HashMap<>();
```

eckter commented on core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/ImpossibleSimulationError.java:
```diff
@@ -12,8 +12,8 @@ public class ImpossibleSimulationError extends OSRDError {
-                "The requested train stopped before reaching its destination (not enough traction)",
-                ErrorCause.USER
+            "The requested train stopped before reaching its destination (not enough traction)",
+            ErrorCause.USER
```
This change causes a checkstyle error, at least locally for me

flomonster reviewed (dismissed):
LGTM

eckter commented on core/src/main/java/fr/sncf/osrd/infra/implementation/tracks/undirected/UndirectedInfraBuilder.java:
```diff
@@ -121,12 +122,15 @@ private void addRemainingLinks(RJSInfra infra) {
-                ));
+                    )
+                )
+                .withContext(new ErrorContext.TrackLink(link.id, srcID, dstID));
```
That's not really how the context is meant to be used, it's generally more useful when catching an error. Here it could be part of the error itself.

I think I just realized what the new model implied. I'm not sure this is the right way to do this... The context is meant to give scoped context to errors that happen that may happen when computing something. This is static data that is already accessible when throwing the error, it should be in the constructor.

eckter commented:
I think the ambiguity comes from the `context`: it doesn't mean the same thing in the core code and in the new error format. But those aren't incompatible either, it just means the error is one level "deeper" in the response dict.

The `context` field in the error should be the serialized `OSRDError`. It should remain mostly unchanged. It still includes the `trace` field, as it was before (a list), giving more context to the error.

The other fields, such as the error type or message, should be "extracted" from the `OSRDError` (and should be removed from its serialization).

Most of the changes should be located in `ExceptionHandler.java` and/or in the serializer of `OSRDError`. We don't change the content of the errors or how they're constructed, just how it's formatted.  

For example: if we throw an `AllowanceConvergenceException`, then catch it to add allowance context. The response used to look (roughly) like this:

```json
{
    "osrd_error_type": "allowance_convergence",
    "convergence_error_type": "DISCONTINUITY",
    "message": "Failed to converge when computing [...]",
    "trace": [
        {
            "type": "allowance",
            "index": 1
        }
    ]
}
```

The new format would look like this:

```json
{
    "status": "failure",
    "warnings": [],
    "error": {
        "type": "allowance_convergence",
        "message": "Failed to converge when computing [...]",
        "context": {
            "convergence_error_type": "DISCONTINUITY"
            "trace": [
                {
                    "type": "allowance",
                    "index": 1
                }
            ]
        }
    }

}
```

Note: most of `InvalidInfraErrors` deserve their own error types for added fields, but I don't think this should be included in the scope of this PR.

anisometropie committed:
core: new error format

ImpossibleSimulationError

Allowance convergence errors

Not implemented

Diagnostic & Strict warnings

AssertWrapper

InfraLoading

pathfinding errors

Discontinuous route error

InvalidInfraDiscontinuousRoute,
InvalidInfraMissingDetectorsRoute, InvalidInfraEndpointAlreadyLinked,

fix

SignalizationError

StandaloneSimulationError

InvalidRollingStock

missing/invalid rolling stock fields

complete invalid rolling stock

invalid schedule

InvalidPathError

unknownRollingStockError

unknownRoute

UnknownTrackSection

missing attribute error

invalid succession

adapter

STDCM error

envelope error

The code is more what I'd call guidelines than actual rules

BAL BARP black sheep...

SigSchema

reservations

envelope

fix

fixes

remove String Format

fixes

style

import

fixes

%s

fix style

fix messed up assert

fix tests

spotbugs warning

javadoc

style

Allowance & infra exceptions

fixes

typos/invalid schedule

reflection

fix allowance tests

refacto invalid infra errors

Khoyo commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/OSRDError.java:
```diff
@@ -68,6 +75,58 @@ public String getMessage() {
+        Map.ofEntries(entry(OSRDError.class, new Adapter())),
+        "osrdErrorType"
+    );
 }
+
+
```
Spurious newlines

flomonster commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/OSRDError.java:
```diff
@@ -25,25 +30,27 @@
+public class OSRDError extends RuntimeException {
 
     @Serial
     private static final long serialVersionUID = 1197516372515951853L;
 
-    public static final String osrdErrorType = "core";
+    public static final transient String osrdErrorType = "core";
 
     /** Detailed error message */
-    public final String message;
+    public transient String message;
 
     /** Stack trace of context */
     public final List<ErrorContext> trace = new ArrayList<>();
 
     /** Whether this is an internal or user error, used to determine if we send a 400 or 500 code */
     public final ErrorCause cause;
```
We could adapt OSRDError to match the serialized type, meaning:

- Three fields: `type`, `message` and `context`.
- The context can contain a `trace` by default.

By doing so we don't have to implement a custom moshi adapter to serialize our errors. 
We don't have to serialize the `cause` (it's only here to define the response code).

Khoyo reviewed (commented):
I'd rather we review the whole design of our error reporting system to something sensible rather than bolt on a custom serializer on it. But failing that, this looks okay.

flomonster reviewed (changes requested):
With this modification, all context information is contained in a `trace` list. Of course, we serialize the errors following the schema, but the semantics are the same as before.

eckter commented:
@flomonster 

> all context information is contained in a trace list

There seems to be a misunderstanding about what this `trace` is. Most of the context isn't in the trace, it's directly located in the (polymorphic) error type. The trace is optional, it's only used to add info we don't have access to *when building the error* to give extra context.

In the current state, none of the infra-related errors have `trace` elements, the IDs of the involved objects are part of the type. It's only used to report allowance errors, to specify the train ID and allowance, on top of other "static" elements that are part of the error itself.

eckter commented:
The branch needs to be rebased, there's extra changes in the diff and conflicting files

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/OSRDError.java:
```diff
@@ -1,15 +1,20 @@
+import java.util.stream.Collectors;
 
-/** Base exception class for the OSRD project. All exceptions in the project should inherit from this.
+/**
+ * Base exception class for the OSRD project. All exceptions in the project should inherit from this.
  *
- * <p>Any type inheriting from this should have a `public static final String osrdErrorType;`
+ * <p>Any type inheriting from this should have a {@code public static final String osrdErrorType;}
  * with a short code describing the error. It is hierarchical, it is concatenated to the one of the parent class.
  *
  * <p>When an exception is caught, context can be added to it. A context stack will be included in the error report.
```
The comment is outdated, there's no inheritance or `withContext`

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/ErrorType.java:
```diff
@@ -0,0 +1,354 @@
+            "Warning was reported with strict mode enabled",
+            ErrorCause.USER
+    ),
+    Assertation(
```
```suggestion
    AssertionError(
```

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/ErrorType.java:
```diff
@@ -0,0 +1,354 @@
+            ErrorCause.USER
+    ),
+    InvalidInfraEndpointAlreadyLinked(
+            "invalid_infra:enpoint_already_linked",
```
```suggestion
            "invalid_infra:endpoint_already_linked",
```

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/ErrorType.java:
```diff
@@ -0,0 +1,354 @@
+            ErrorCause.USER
+    ),
+    PathfindingElectrificationError(
+            "no_path_found:eletrification",
```
```suggestion
            "no_path_found:electrification",
```

eckter commented on core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/AllowanceRangesTests.java:
```diff
@@ -320,17 +326,19 @@ public void regressionTestCornerCase() {
+                () -> allowance.apply(maxEffortEnvelope, testContext)
+        );
+        assertEquals(err.osrdErrorType, ErrorType.AllowanceConvergenceTooMuchTime);
     }
 
     /** Applies the allowance to the envelope. Any user error (impossible margin) is ignored */
     private void applyAllowanceIgnoringUserError(Allowance allowance, Envelope envelope, EnvelopeSimContext context) {
         try {
             allowance.apply(envelope, context);
-        } catch (AllowanceConvergenceException e) {
-            assertEquals(OSRDError.ErrorCause.USER, e.cause);
+        } catch (OSRDError e) {
+            assertEquals(e.cause, ErrorCause.USER);
```
This catches more errors than before, it should only ignore the allowance related errors

anisometropie replied:
not exactly sure how to do that

anisometropie replied:
made the change, 
```
            if (
                    e.osrdErrorType == ErrorType.AllowanceConvergenceDiscontinuity
                    || e.osrdErrorType == ErrorType.AllowanceConvergenceNotEnoughTime
                    || e.osrdErrorType == ErrorType.AllowanceConvergenceTooMuchTime
            ) {
                assertEquals(e.cause, ErrorCause.USER);
            }
```

eckter replied:
That would ignore the rest of the errors.

The previous behavior was catching allowance errors and ignoring only the user-based ones, while not catching the other errors (which would make the test fail). Now that it's based on `ErrorType`, the only user-based allowance error is `TooMuchTime`, you should ignore this one and fail if any other error is thrown.

eckter commented on core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/AllowanceTests.java:
```diff
@@ -714,7 +719,7 @@ public void testShortLinear() {
                         new AllowanceRange(800, 810, new AllowanceValue.Percentage(10))
                 )
         );
-        assertThrows(AllowanceConvergenceException.class, () -> allowance.apply(maxEffortEnvelope, testContext));
+        assertThrows(OSRDError.class, () -> allowance.apply(maxEffortEnvelope, testContext));
```
It should be specifically a convergence exception

anisometropie replied:
Added a check for the type
```
        var err = assertThrows(
                OSRDError.class,
                () -> allowance.apply(maxEffortEnvelope, testContext)
        );
        assertEquals(err.osrdErrorType, ErrorType.AllowanceConvergenceTooMuchTime);
```

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/Reflection.java:
```diff
@@ -13,6 +13,7 @@
 import java.util.List;
+import java.util.Map;
 
 public class Reflection {
```
The whole class can be dropped: it's only used to make a polymorphic warning adapter, and the warnings aren't polymorphic anymore

anisometropie replied:
Done

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/ErrorType.java:
```diff
@@ -0,0 +1,354 @@
+            "Infinite loop in signal dependency updates",
+            ErrorCause.USER
+    ),
+    StandaloneSimulationError(
+            "standalone_simulation",
+            "",
+            ErrorCause.INTERNAL
+    ),
```
Empty message and not used in the project

```suggestion
```

anisometropie replied:
done

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/ErrorType.java:
```diff
@@ -0,0 +1,354 @@
+    public final String message;
+    public final ErrorCause cause;
+
+    private ErrorType(String type, String message, ErrorCause cause) {
```
It's implied for enum constructors

```suggestion
    ErrorType(String type, String message, ErrorCause cause) {
```

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/ErrorType.java:
```diff
@@ -0,0 +1,354 @@
+package fr.sncf.osrd.reporting.exceptions;
+
+public enum ErrorType {
```
You have all my respect for writing all that, thanks a lot

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/OSRDError.java:
```diff
@@ -24,50 +29,554 @@
+     * @return a new OSRDError instance
+     */
+    public static OSRDError newInvalidScheduleError(ErrorType errorType) {
+        return new OSRDError(errorType);
+    }
+
+    /**
+     * Creates a new OSRDError for an invalid schedule error with a track section.
+     *
+     * @param errorType     the error type
+     * @param trackSection  the track section associated with the error
+     * @return a new OSRDError instance
+     */
+    public static OSRDError newInvalidScheduleError(
+            ErrorType errorType,
+            String trackSection
+    ) {
+        var error = new OSRDError(errorType);
+        error.context.put("track_section", trackSection);
+        return error;
+    }
+
+    /**
+     * Creates a new OSRDError for an invalid schedule error with a route and route signaling type.
+     *
+     * @param errorType           the error type
+     * @param route               the route associated with the error
+     * @param routeSignalingType  the route signaling type
+     * @return a new OSRDError instance
+     */
+    public static OSRDError newInvalidScheduleError(
+            ErrorType errorType,
+            String route,
+            String routeSignalingType
+    ) {
+        var error = new OSRDError(errorType);
+        error.context.put("route", route);
+        error.context.put("route_signaling_type", routeSignalingType);
+        return error;
+    }
```
All these overloads feel weird, especially the first one that does nothing but forwarding to the constructor.

Maybe there could be some cases where we can call the constructor directly, maybe with an extra `context` parameter to avoid a high number of overloads. 

This also applies to the invalid infra errors

eckter commented on core/osrd-reporting/src/main/java/fr/sncf/osrd/reporting/exceptions/OSRDError.java:
```diff
@@ -24,50 +29,554 @@
+        adapter = moshi.adapter(OSRDError.class);
+    }
+
+    @Serial
+    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
+        throw new RuntimeException("Deserialization isn't supported for OSRDError");
+    }
```
A comment would be nice to explain why it's needed

anisometropie replied:
done

flomonster reviewed (approved):
Looks good (just reviewed not tested)
